src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -1086,36 +1203,47 @@ class AutoDiffSort
-        SortOp::attachInterface<ADDataFlowSortOp>(*context);
-        ScatterOp::attachInterface<ADDataFlowScatterOp>(*context);
-        ReduceOp::attachInterface<ADDataFlowReduceOp>(*context);
-
-        CaseOp::attachInterface<RegionBranchCaseOp>(*context);
-
-        ScatterOp::attachInterface<ScatterActivity>(*context);
-        ScatterOp::attachInterface<AutoDiffScatter>(*context);
-
-        ReturnOp::attachInterface<AutoDiffHLOReturn>(*context);
-
-        ReduceOp::attachInterface<AutoDiffReduceFwd<ReduceOp>>(*context);
-        IfOp::attachInterface<AutoDiffIfRev>(*context);
-        IfOp::attachInterface<AutoDiffIfFwd>(*context);
-        IfOp::attachInterface<AutoDiffIfCF>(*context);
-        WhileOp::attachInterface<AutoDiffWhileFwd>(*context);
-        ReduceOp::attachInterface<AutoDiffReduceCF<ReduceOp>>(*context);
-        WhileOp::attachInterface<AutoDiffReduceCF<WhileOp>>(*context);
-        BroadcastInDimOp::attachInterface<AutoDiffBroadcastInDimRev>(*context);
-        SliceOp::attachInterface<AutoDiffSliceRev>(*context);
-        ReduceOp::attachInterface<AutoDiffReduceRev>(*context);
-        ConcatenateOp::attachInterface<AutoDiffConcatenateRev>(*context);
-        ConstantOp::attachInterface<SHLOConstantOpBatchInterface>(*context);
-        TransposeOp::attachInterface<SHLOTransposeOpBatchInterface>(*context);
-      });
+  registry.addExtension(+[](MLIRContext *context,
+                            stablehlo::StablehloDialect *) {
+    registerInterfaces(context);
+
+    // SortOp::attachInterface<AutoDiffSort>(*context);
+
+    WhileOp::attachInterface<ADDataFlowWhileOp>(*context);
+    SortOp::attachInterface<ADDataFlowSortOp>(*context);
+    ScatterOp::attachInterface<ADDataFlowScatterOp>(*context);
+    ReduceOp::attachInterface<ADDataFlowReduceOp>(*context);
+
+    CaseOp::attachInterface<RegionBranchCaseOp>(*context);
+
+    ScatterOp::attachInterface<ScatterActivity>(*context);
+    ScatterOp::attachInterface<AutoDiffScatter>(*context);
+
+    ReturnOp::attachInterface<AutoDiffHLOReturn>(*context);
+
+    ReduceOp::attachInterface<AutoDiffReduceFwd<ReduceOp>>(*context);
+    IfOp::attachInterface<AutoDiffIfRev>(*context);
+    IfOp::attachInterface<AutoDiffIfFwd>(*context);
+    IfOp::attachInterface<AutoDiffIfCF>(*context);
+
+    WhileOp::attachInterface<AutoDiffWhileFwd>(*context);
+    ReduceOp::attachInterface<AutoDiffReduceCF<ReduceOp>>(*context);
+    WhileOp::attachInterface<AutoDiffReduceCF<WhileOp>>(*context);
+    BroadcastInDimOp::attachInterface<AutoDiffBroadcastInDimRev>(*context);
+    SliceOp::attachInterface<AutoDiffSliceRev>(*context);
+    ReduceOp::attachInterface<AutoDiffReduceRev>(*context);
+    ConcatenateOp::attachInterface<AutoDiffConcatenateRev>(*context);
+
+    ConstantOp::attachInterface<SHLOConstantOpBatchInterface>(*context);
+    TransposeOp::attachInterface<SHLOTransposeOpBatchInterface>(*context);
+    IfOp::attachInterface<SHLOUnrollOpBatchInterface<IfOp>>(*context);
```

Comment:
hm the control flow ops should definitely not use the interface here at a significant perf loss. Since effectively it puts the batching outside the if/while by unrolling, whereas putting it on the inside retains the ability to fuse/vectorize/etc

Reply:
yeah, it is not ideal as it generates a lot of code (proportional to the number of samples in the batch) but for IfOp and WhileOp it cannot know whether or not the same branch/number of iterations will be taken for all samples. I will look at how xla handles `stablehlo.map` under the hood. Or maybe we can emit the op inside a `stablehlo.map` instead.

Reply:
you can tho! check if it's a constant which is a splat [see examples in enzymehlo-opt], which means its the same across all iters, or a broadcast from a single value